### ISSUE TITLE
Evaluate default values when a function is called, not at compile time

### ISSUE THREAD

**Issue by hanche** [elves/elvish#998]
Currently, default values are apparently evaluated when a function is compiled, not when it is evaluated:

```
⬥ x = 1
⬥ fn f [&v=$x]{ put $v }
⬥ x = 2
⬥ f
⮕ 1
```

A reasonable alternative is instead to evaluate at call time, in which case the above example would produce the value `2`. This could usefully be employed in many ways, as for example supplying `(date '+%F %T')` as the default for a time stamp, which the caller could override at will. Of course, this is easily accomplished otherwise, for example by letting the default be `$nil` and having the function check for this and computing a different value in that case: But this requires more coding.

If the behaviour is changed as proposed, then if you really want to compute a default value at compile time, you would need to work harder, for example saying

```
⬥ f~ = ({ local:vdef = $x; put [&v=$vdef]{ put $v }})
```

Yes, that is ugly, and maybe better workarounds can be found. But I think the need will be minimal or even nonexistent: Most times, default values will be constants anyway.

Perhaps the strongest argument for the proposal, though: I said the default values are computed at compile time. But that is apparently not the truth, as can be seen by wrapping my initial example in a function call:

```
⬥ fn foo {
    local:x = 1
    fn f { put $x }
    fn g { x=2 f }
    g
  }
⬥ foo
⮕ 2
```

It is confusing that the same code produces different results depending on whether it's at top level or in a function body.

I haven't surveyed a great number of languages, but javascript and common lisp both evaluate default arguments at call time, whereas python evaluates them at compile time.

**Comment by krader1961** [comment]
As @hanche noted, Python also evaluates the default value of a function argument at compile time. This is a well known gotcha. See, for example, https://docs.python-guide.org/writing/gotchas/. The solution is to use a value such as `None` that would otherwise not make any sense and in the body of the function check for that magic value and substitute the desired default. In the context of Elvish that would mean doing something like this:

```
x = 1
fn f [&v=$nil]{
       if (is $v $nil) {
              v = $x
        }
         put $v
}
f
▶ 1
x = 2
f
▶ 2
```

@hanche's second example is quite different in structure from the first as it uses a closure rather than a function argument default value. I think that's an apples to oranges comparison.

My opinion is likely biased by having programmed in Python for thirteen years but I vote for retaining the current behavior. It simply needs to be clearly documented, including an example or two for how to deal with defaults that are not immutable constants.

**Comment by hanche** [comment]
@krader1961 Regarding my example: Sorry, I don't know what bit me. The test I had intended to run does indeed shows the same behaviour in the closure case as at top level. It looks like this:

```
⬥ fn foo {
    local:x = 1
    fn f [&v=$x]{ put $v }
    fn g { x=2 f}
    g }
⬥ foo
⮕ 1
```

I'd probably write the test in your example more briefly as `v = (or $v $x)` myself, unless of course I need `$false` to be a viable value for the option.

I admit I have no stronger arguments for my proposal than “it's what I am used to from CL”, and that I would find it more convenient for the most part. If you can't come up with a stronger argument than it being what you are used to, from Python, well I admit it is a bit stronger than mine, since there are more Python users than Lisp users in the world. OTOH, if @xiaq wants elvish to be somewhat Lisp-like, then evaluation at function call time seems the right option. In either case, he gets to have the last word on this.

My main reason for raising the issue is to ensure that we don't freeze the current behaviour without at least having thought it through.

And _definitely_, whatever the end result is, it needs to be documented properly.

**Comment by hanche** [comment]
I don't know that it matters, but I was wrong to imply that default values are assigned at compile time. Rather, they seem to be assigned when the `fn` form is executed:

```
⬥ x = 1
⬥ fn foo {
    fn f [&v=$x]{ put $v }
    x = 3
    f
  }
⬥ x = 2
⬥ foo
⮕ 2
```
If the default value were assigned at compile time, I'd have expected the result to be 1, not 2.

**Comment by krader1961** [comment]
@hanche, Your most recent example is, again, due to closure behavior. It behaves the same way, more or less, in Python. See below. You are correct that the value is not bound at "compile" time but when the function is defined. Which, if the function definition isn't nested is at "compile" time :smile:

```python
x = 1

def foo():
    global x
    def subfoo(y=x):
        print(y)
    x = 666
    subfoo()

foo()
x = 2
foo()
```

**Comment by hanche** [comment]
@krader1961 Yes, I think you summarised what I tried to say in different words.

@xiaq When you get around to this issue, whether you are going to close it, slap a Maybe tag on it, or whatever, could you change “compile time” to “function definition time” in the title?

**Comment by xiaq** [comment]
Yeah, Elvish mimics Python instead of JavaScript and Common Lisp here. I'm not sure now, but slightly in favor of changing the evaluation to be done at call time.

The [most notorious bug](https://lethain.com/digg-v4/) caused by Python's behavior isn't actually possible for Elvish since Elvish lists and maps are value types, not reference types, so you can't mutate the default value.

**Comment by xiaq** [comment]
The major downside of evaluating default values at call-time is that it makes introspection much harder; `$func[opt-defaults][$opt-name]` can no longer give you an actual value. In fact, since the default value can depend on the value of *other arguments and options* (if we were to imitate JavaScript), it needs to be a function that takes all other arguments and options.
